protocols/stream/src/behaviour.rs

```diff
@@ -0,0 +1,287 @@
+use core::fmt;
+use std::{
+    collections::{HashMap, VecDeque},
+    io,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use futures::{channel::mpsc, StreamExt};
+use libp2p_core::{Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    self as swarm, dial_opts::DialOpts, ConnectionDenied, ConnectionId, FromSwarm,
+    NetworkBehaviour, Stream, StreamProtocol, THandler, THandlerInEvent, THandlerOutEvent, ToSwarm,
+};
+use rand::seq::IteratorRandom;
+use swarm::{
+    behaviour::ConnectionEstablished, dial_opts::PeerCondition, ConnectionClosed, DialError,
+    DialFailure,
+};
+
+use crate::{
+    handler::{Handler, NewStream, RegisterProtocol},
+    Control, IncomingStreams,
+};
+
+/// A generic behaviour for stream-oriented protocols.
+pub struct Behaviour {
+    shared: Arc<Mutex<Shared>>,
+    receiver: mpsc::Receiver<PeerId>,
+    supported_protocols: HashMap<StreamProtocol, mpsc::Sender<(PeerId, Stream)>>,
+
+    events: VecDeque<ToSwarm<(), RegisterProtocol>>,
+}
+
+pub(crate) struct Shared {
+    connections: HashMap<ConnectionId, PeerId>,
+    senders: HashMap<ConnectionId, mpsc::Sender<NewStream>>,
+
+    pending_channels: HashMap<PeerId, (mpsc::Sender<NewStream>, mpsc::Receiver<NewStream>)>,
+
+    dial_sender: mpsc::Sender<PeerId>,
+}
+
+impl Shared {
+    pub(crate) fn new(dial_sender: mpsc::Sender<PeerId>) -> Self {
+        Self {
+            connections: Default::default(),
+            senders: Default::default(),
+            pending_channels: Default::default(),
+            dial_sender,
+        }
+    }
+
+    pub(crate) fn sender(&mut self, peer: PeerId) -> mpsc::Sender<NewStream> {
+        let maybe_sender = self
+            .connections
+            .iter()
+            .filter_map(|(c, p)| (p == &peer).then_some(c))
+            .choose(&mut rand::thread_rng())
+            .and_then(|c| self.senders.get(c));
+
+        match maybe_sender {
+            Some(sender) => {
+                tracing::debug!("Returning sender to existing connection");
+
+                sender.clone()
```

Comment:
This will expand the size of the channel and thus break backpressure, no?

Reply:
Yes but a user would have to to out of their way to achieve that by also cloning the `Control`.

If a user calls `open_stream` in a loop, we will still uphold backpressure and only open one stream at a time.
If they clone the `Control` or don't immediately await the future then a new `Sender` will be created for each one. I think that is fairly intuitive actually. I think the only concern here is increased memory-consumption and doing any of the above two suggests to the user that more memory will be consumed.

Overall, I think this is acceptable. Do you see any other problems?

Reply:
Even awaiting multiple futures doesn't work because we take `&mut self` so the user MUST clone the `Sender`. In other words, we are merely passing on the behaviour of an `mpsc::Sender` to `Control`.

We should probably document that though.